sno/working_copy/sqlserver.py

```diff
@@ -0,0 +1,553 @@
+import contextlib
+import logging
+import re
+import time
+from urllib.parse import urlsplit, urlunsplit
+
+import click
+import sqlalchemy
+from sqlalchemy import literal_column
+from sqlalchemy.ext.compiler import compiles
+from sqlalchemy.orm import sessionmaker
+from sqlalchemy.sql import crud, quoted_name
+from sqlalchemy.sql.dml import ValuesBase
+from sqlalchemy.sql.functions import Function
+from sqlalchemy.sql.compiler import IdentifierPreparer
+from sqlalchemy.types import UserDefinedType
+
+from .base import WorkingCopy
+from . import sqlserver_adapter
+from .table_defs import SqlServerSnoTables
+from sno import crs_util
+from sno.geometry import Geometry
+from sno.exceptions import InvalidOperation
+from sno.sqlalchemy import sqlserver_engine
+
+
+"""
+* database needs to exist
+* database needs to have postgis enabled
+* database user needs to be able to:
+    1. create 'sno' schema & tables
+    2. create & alter tables in the default (or specified) schema
+    3. create triggers
+"""
+
+L = logging.getLogger("sno.working_copy.postgis")
+
+
+class WorkingCopy_SqlServer(WorkingCopy):
+    def __init__(self, repo, uri):
+        """
+        uri: connection string of the form mssql://[user[:password]@][netloc][:port][/dbname/schema][?param1=value1&...]
+        """
+        self.L = logging.getLogger(self.__class__.__qualname__)
+
+        self.repo = repo
+        self.uri = uri
+        self.path = uri
+
+        url = urlsplit(uri)
+
+        if url.scheme != "mssql":
+            raise ValueError("Expecting mssql://")
+
+        url_path = url.path
+        path_parts = url_path[1:].split("/", 3) if url_path else []
+        if len(path_parts) != 2:
+            raise ValueError("Expecting mssql://[HOST]/DBNAME/SCHEMA")
+        url_path = f"/{path_parts[0]}"
+        self.db_schema = path_parts[1]
+
+        # Rebuild DB URL suitable for sqlserver_engine.
+        self.dburl = urlunsplit([url.scheme, url.netloc, url_path, url.query, ""])
+        self.engine = sqlserver_engine(self.dburl)
+        self.sessionmaker = sessionmaker(bind=self.engine)
+        self.preparer = IdentifierPreparer(self.engine.dialect)
+
+        self.sno_tables = SqlServerSnoTables(self.db_schema)
+
+    @classmethod
+    def check_valid_creation_path(cls, workdir_path, path):
+        # TODO - promote to superclass
```

Comment:
Done - added a superclass with code common to SQL Server and PostGIS - 
DatabaseServer_WorkingCopy.
All the DB URI code is moved to there since its common to both.